src/array/reducer/swap-control.ts

```diff
@@ -0,0 +1,52 @@
+import { Actions, SwapArrayControlAction } from '../../actions';
+import { computeArrayState, FormArrayState } from '../../state';
+import { childReducer, updateIdRecursive } from './util';
+
+function swapArrayValues(a: ReadonlyArray<any>, i: number, j: number) {
+  const n = [ ...a ];
+  [ n[ i ], n[ j ] ] = [ n[ j ], n[ i ] ];
+  return n;
+}
+
+export function swapControlReducer<TValue>(
+  state: FormArrayState<TValue>,
+  action: Actions<TValue[]>,
+): FormArrayState<TValue> {
+  if (action.type !== SwapArrayControlAction.TYPE) {
+    return state;
+  }
+
+  if (action.controlId !== state.id) {
+    return childReducer(state, action);
+  }
+
+  const from = action.from;
+  const to = action.to;
+
+  if (from === to) {
+    return state;
+  }
+
+  if (from >= state.controls.length || to >= state.controls.length) {
+    throw new Error(`Index [${from >= state.controls.length ? `from:${from}` : `to:${to}`}]
+     is out of bounds for array '${state.id}' with length ${state.controls.length}!`);
+  }
+
+  let controls = swapArrayValues(state.controls, from, to);
+  controls = controls.map((c, i) => (i >= from || i >= to) ? updateIdRecursive(c, `${state.id}.${i}`) : c);
```

Comment:
While the additional guard for checking whether the ID needs to be updated is correct it is redundant since inside the `updateIdRecursive` function the ID is also compared to the previous ID and the state is returned unmodified if it is the same. Therefore you can reduce the cyclomatic complexity of this code by simply omitting the check.